x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```diff
@@ -0,0 +1,189 @@
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.action.update.UpdateResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.search.action.AsyncSearchResponse;
+import org.elasticsearch.xpack.core.search.action.GetAsyncSearchAction;
+import org.elasticsearch.xpack.core.search.action.SubmitAsyncSearchAction;
+import org.elasticsearch.xpack.core.search.action.SubmitAsyncSearchRequest;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+public class TransportSubmitAsyncSearchAction extends HandledTransportAction<SubmitAsyncSearchRequest, AsyncSearchResponse> {
+    private static final Logger logger = LogManager.getLogger(TransportSubmitAsyncSearchAction.class);
+
+    private final NodeClient nodeClient;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Supplier<InternalAggregation.ReduceContext> reduceContextSupplier;
+    private final TransportSearchAction searchAction;
+    private final AsyncSearchStoreService store;
+    private final Random random;
+
+    @Inject
+    public TransportSubmitAsyncSearchAction(ClusterService clusterService,
+                                            TransportService transportService,
+                                            ActionFilters actionFilters,
+                                            NamedWriteableRegistry registry,
+                                            Client client,
+                                            NodeClient nodeClient,
+                                            SearchService searchService,
+                                            TransportSearchAction searchAction) {
+        super(SubmitAsyncSearchAction.NAME, transportService, actionFilters, SubmitAsyncSearchRequest::new);
+        this.clusterService = clusterService;
+        this.threadPool = transportService.getThreadPool();
+        this.nodeClient = nodeClient;
+        this.reduceContextSupplier = () -> searchService.createReduceContext(true);
+        this.searchAction = searchAction;
+        this.store = new AsyncSearchStoreService(taskManager, threadPool, client, registry);
+        this.random = new Random(System.nanoTime());
```

Comment:
This looks dangerous to me since it would be legal for System#nanoTime to have a resolution of tens of milliseconds, so two requests could get the same seed. I wonder why you don't call `UUIDs.randomBase64UUID()` (without a Random parameter) below, which uses a shared secure random?

Reply:
+1 thanks, I moved to `UUIDs.randomBase64UUID()` in https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11